bool fl
bool fr
bool hl
bool hr
